lib/canonical_version.sh

```diff
@@ -1,4 +1,28 @@
-canonicalVersion() {
+#!/usr/bin/env bash
+
+fetch_erlang_versions() {
+  url="https://raw.githubusercontent.com/HashNuke/heroku-buildpack-elixir-otp-builds/master/otp-versions"
+  curl -s "$url" 
+}
+
+exact_erlang_version_available() {
+  version=$1
+  available_versions=$2
+  found=1
+  while read -r line; do
+    if [ "$line" = "$version" ]; then
+      found=0
```

Comment:
it's probably better to use `return 0` here

Reply:
ping @jesseshieh

Reply:
I tried to use `return` at first but shunit2 doesn't play as nicely with it. It kinda sucks that the test would affect the code, but I really wanted tests and there aren't many bash test frameworks available.

Reply:
return is standard bash, what about it doesn't play nicely?

Reply:
also I don't see any tests written

Reply:
without any tests to go off of for what you've already wrote testing return instead of echo with shunit2 should be as easy as: 
```
test_included_version_is_valid() {
  if ! (exact_erlang_version_available "$valid_version" "$(fetch_erlang_versions)"); then
    fail "existing version was not found"
  fi
}
```

Reply:
and:
```
test_missing_version_is_not_valid() {
  if (exact_erlang_version_available "$invalid_version" "$(fetch_erlang_versions)"); then
    fail "missing version was found"
  fi
}
```

Reply:
nvmd found the test files. with return you can still do:
```
testExactErlangVersionAvailable() {
  assertFalse (exact_erlang_version_available "22.0.3" "$VERSIONS")
}

testExactErlangVersionUnavailable() {
  assertTrue (exact_erlang_version_available "22.0.1" "$VERSIONS")
}

testExactErlangVersionMajorOnly() {
  assertTrue (exact_erlang_version_available "21" "$VERSIONS")
}

testExactErlangVersionMajorMinorOnly() {
  assertFalse (exact_erlang_version_available "21.0" "$VERSIONS")
}
```

Reply:
Oh, the tests are in the file named canonical_version_test.sh. Here's a link to it 
https://github.com/HashNuke/heroku-buildpack-elixir/pull/158/files#diff-9cc02e1022783a7b53a2d6aa8e80994f

I'm honestly okay with doing it your way, but I'd like to defer to @HashNuke if possible. To me, there isn't really a clear correct option. Return codes are "built-in" but you can only return integers so for functions that need to "return" strings, you have to use `echo` anyway. If some functions use `echo` and some use `return`, the inconsistency can get a little confusing.

For shunit2, the current tests are of the form
```
assertEquals 0 my_function
```

If we switch from `echo` to `return`, it'll change to the following, which isn't a big deal, but feels a little harder to read to me.
```
my_function
result=$?
assertEqual 0 $result
```

Reply:
I tried using `assertTrue` and `assertFalse` earlier, but it doesn't seem to work easily with `return`. Here is the source code for it.
https://github.com/kward/shunit2/blob/ba130d69bbff304c0c6a9c5e8ab549ae140d6225/shunit2#L442

It seems to work on
1. a command that `echo`s 0 or 1
2. a condition such as `[ 0 -eq 0 ]`

Reply:
If the code I posted above doesn't work in either form (it is a condition) then keep it echoed, I just don't like the idea of using echo instead of the built in way script statuses are meant to be used